### DISCUSSION TITLE
Looking for guidance on Character/CharacterVirtual setup

### DISCUSSION THREAD

**Discussion by Froyok** [amerkoleci/joltc#3] [Q&A]
Hi,

I have been using this library/wrapper in my project (thx for making it by the way !) and today I wanted to setup a character (to test walking around a scene).
I got basic static walls and falling spheres/cubes working, but I can't seem to figure out how to get a Character to update (at minimum to fall under the scene/world gravity).

I looked at the native Jolt C++ sample to see how to setup a character, but the CSharp/C API version is quite different. I'm likely missing something but so far I haven't found what.

This is my pseudo code:

```lua
-- Character creation
	CharacterUpdateSettings = New( "JPH_ExtendedUpdateSettings" )
	CharacterUpdateSettings.stickToFloorStepDown = JPH_Vec3( 0, 0, 0 )
	CharacterUpdateSettings.walkStairsStepUp = JPH_Vec3( 0, 0, 0 )

	Position = JPH_RVec3( 0, 2, 0 )
	Rotation = JPH_Quat( 0, 0, 0, 1 )

	Settings = JPH_CharacterVirtualSettings_Create()
	BaseSettings = Cast( "JPH_CharacterBaseSettings*", Settings )
	Shape = JPH_CapsuleShape_Create(
		1.8 / 2.0, -- Half of 1.8m tall
		0.4
	)
	JPH_CharacterBaseSettings_SetShape(
		BaseSettings,
		Cast( "JPH_Shape*", Shape )
	)
	JPH_CharacterBaseSettings_SetUp(
		BaseSettings,
		JPH_Vec3( 0, 1, 0 )
	)

	Character = PHYSICS_C.JPH_CharacterVirtual_Create(
		Settings,
		Position,
		Rotation,
		0,
		PHYSICS.System
	)

-- Update code:
	JPH_CharacterVirtual_ExtendedUpdate(
		Character,
		DeltaTime,
		CharacterUpdateSettings,
		PHYSICS.Layers.Object.Dynamic, -- aka Layers::MOVING
		PHYSICS.System
	)

	JPH_PhysicsSystem_Step(
		PHYSICS.System,
		DeltaTime,
		2
	)
```

With that code, if I retrieve the character position, it is static (always at the one given at creation). I called `SetLinearVelocity()` with an up vector (`(0, 1, 0)`) to simulate a push but then the Character goes up infinitely.

So how can I simulate gravity on that character ?

**Comment by amerkoleci** [comment]
Hi,
First of all can you show some jolt C++ code that does the same?

After I can try to reproduce it and see if there is some bindings missing. 

Thanks

**Reply by Froyok** [reply]
This is the code I have been mostly looking at: https://github.com/jrouwe/JoltPhysics/blob/master/Samples/Tests/Character/CharacterVirtualTest.cpp
There are some bits inherited from there as well: https://github.com/jrouwe/JoltPhysics/blob/master/Samples/Tests/Character/CharacterBaseTest.cpp

**Reply by Froyok** [reply]
To clarify: I mostly looked at `CharacterVirtualTest::Initialize()` for the settings and `CharacterBaseTest::PrePhysicsUpdate()` (which calls Character's `ExtendedUpdate()`).

**Reply by Froyok** [reply]
Okay it looks like you are supposed to apply gravity yourself on Characters, has seen in: https://github.com/jrouwe/JoltPhysics/blob/master/Samples/Tests/Character/CharacterBaseTest.cpp#L647

**Reply by Froyok** [reply]
Looks like I got my answer, sorry for wasting your time.

**Comment by Froyok** [comment]
I took the time to dig a bit more into the examples and the source code of jolt.
I stumbled on this comment from Jolt source itself:

> Runtime character object.
> This object usually represents the player. Contrary to the Character class it doesn't use a rigid body but moves doing collision checks only (hence the name virtual). The advantage of this is that you can determine when the character moves in the frame (usually this has to happen at a very particular point in the frame) but the downside is that other objects don't see this virtual character. In order to make this work it is recommended to pair a CharacterVirtual with a Character that moves along. This Character should be keyframed (or at least have no gravity) and move along with the CharacterVirtual so that other rigid bodies can collide with it.

From: [CharacterVirtual.h](https://github.com/jrouwe/JoltPhysics/blob/2292c415a9a94cd7b4706de6ac8f24c9b5af74a5/Jolt/Physics/Character/CharacterVirtual.h#L159)

I'm wondering if the Character class/functions should be exposed as well ?
That could make it easier to follow the samples and replicate some of their behaviors.

**Comment by amerkoleci** [comment]
Just added Character support (WIP) https://github.com/amerkoleci/joltc/commit/04fc220dcf30e60ef3e5e5395d1e66e9d4942e79

Tell me what you think

**Reply by Froyok** [reply]
Thx, I'm going to check that out ! :)

**Reply by Froyok** [reply]
Sorry for not replying sooner, I needed to get more familiar with Jolt first.
So while Jolt code recommend using an intermediate Character, the samples use instead a simpler approach. So in the end it seems that what I'm missing are two properties from the CharacterVirtualSettings struct:

```c++
	/* 
	This character can optionally have an inner rigid body. 
	This rigid body can be used to give the character presence 
	in the world. When set it means that:

	- Regular collision checks (e.g. NarrowPhaseQuery::CastRay) 
	will collide with the rigid body (they cannot collide with CharacterVirtual 
	since it is not added to the broad phase)

	- Regular contact callbacks will be called through the ContactListener 
	(next to the ones that will be passed to the CharacterContactListener)

	- Fast moving objects of motion quality LinearCast will not be 
	able to pass through the CharacterVirtual in 1 time step
	*/
	RefConst<Shape>						mInnerBodyShape;	

	/// Layer that the inner rigid body will be added to
	ObjectLayer							mInnerBodyLayer = 0;
```
([From here](https://github.com/jrouwe/JoltPhysics/blob/a8327c3ff83ec1792970dc9fc29bf23cef327ba6/Jolt/Physics/Character/CharacterVirtual.h#L47).)
You can see it used [in the samples here](https://github.com/jrouwe/JoltPhysics/blob/a8327c3ff83ec1792970dc9fc29bf23cef327ba6/Samples/Tests/Character/CharacterVirtualTest.cpp#L34).

In the current state of the repo I only saw all the properties that come before those two.

Meanwhile, thank you for taking to time to expose the Character class. That's going to be useful for other cases.
